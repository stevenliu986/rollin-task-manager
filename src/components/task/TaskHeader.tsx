import { Button } from "antd";
import leadingIcon from '../../assets/images/Leading Icon.png'
import type { ITask } from "./AddTask";

interface ITaskHeaderProps {
    taskDetail: ITask;
}

export default function TaskHeader({ taskDetail }: Readonly<ITaskHeaderProps>) {
    const due = taskDetail.due?.slice(0, -5)

    return (<div style={{ display: 'flex', justifyContent: 'space-between', marginBottom: 16 }}>

        <Button
            type="default"
            size="small"
            style={{
                backgroundColor: taskDetail.complete ? '#11a436' : '#196BE5',
                border: `1px solid ${taskDetail.complete ? '#11a436' : '#196BE5'}`,
                borderRadius: '25px',
                padding: '0 8px',
                display: 'flex',
                alignItems: 'center',
            }}
        >
            <span style={{ fontSize: '0.875rem', color: '#FFF' }}>
                {taskDetail.complete ? 'Completed' : 'In Progress'}
            </span>
        </Button>
        <Button
            type="default"
            size="small"
            style={{
                backgroundColor: '#F5F5F5',
                borderRadius: '25px',
                padding: '0 8px',
                display: 'flex',
                alignItems: 'center',
            }}
        >
            <img src={leadingIcon} alt="leading-icon" />
            <span style={{ fontSize: '0.875rem' }}>
                {due}
            </span>
        </Button>
    </div>)
}