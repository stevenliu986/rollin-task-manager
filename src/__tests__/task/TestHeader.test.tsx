import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import TaskHeader from '../../components/task/TaskHeader';
import type { ITask } from '../../components/task/AddTask';

vi.mock('../../assets/images/Leading Icon.png', () => ({
    default: 'test-leading-icon.png'
}));

vi.mock('antd', async (importOriginal) => {
    const mod = await importOriginal();
    return {
        ...mod,
        Button: ({ children, style, ...props }: any) => (
            <button style={style} {...props} data-testid="ant-button">
                {children}
            </button>
        )
    };
});

describe('TaskHeader Component', () => {
    const mockTask: ITask = {
        id: 1,
        title: 'Test Task',
        description: 'Test Description',
        complete: false,
        created: '2023-01-01',
        due: '2023-12-31T00:00:00.000Z'
    };

    const mockCompletedTask: ITask = {
        ...mockTask,
        complete: true
    };

    it('renders status button correctly for incomplete task', () => {
        render(<TaskHeader taskDetail={mockTask} />);

        const statusButton = screen.getAllByTestId('ant-button')[0];
        expect(statusButton).toHaveTextContent('In Progress');
        expect(statusButton).toHaveStyle({
            backgroundColor: '#196BE5',
            border: '1px solid #196BE5'
        });
    });

    it('renders status button correctly for completed task', () => {
        render(<TaskHeader taskDetail={mockCompletedTask} />);

        const statusButton = screen.getAllByTestId('ant-button')[0];
        expect(statusButton).toHaveTextContent('Completed');
        expect(statusButton).toHaveStyle({
            backgroundColor: '#11a436',
            border: '1px solid #11a436'
        });
    });

    it('renders due date button when due date exists', () => {
        render(<TaskHeader taskDetail={mockTask} />);

        const dueButton = screen.getAllByTestId('ant-button')[1];
        expect(dueButton).toHaveTextContent('2023-12-31');
        expect(dueButton).toHaveStyle({
            backgroundColor: '#F5F5F5'
        });
        expect(screen.getByRole('img')).toHaveAttribute('alt', 'leading-icon');
    });

    it('formats due date correctly by removing last 5 characters', () => {
        render(<TaskHeader taskDetail={{ ...mockTask, due: '2023-12-31T23:59:59.000Z' }} />);

        const dueButton = screen.getAllByTestId('ant-button')[1];
        expect(dueButton).toHaveTextContent('2023-12-31T23:59:59');
    });

    it('matches snapshot for incomplete task with due date', () => {
        const { asFragment } = render(<TaskHeader taskDetail={mockTask} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot for completed task without due date', () => {
        const { asFragment } = render(<TaskHeader taskDetail={mockCompletedTask} />);
        expect(asFragment()).toMatchSnapshot();
    });
});