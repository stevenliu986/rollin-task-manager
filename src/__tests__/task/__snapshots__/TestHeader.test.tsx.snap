// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskHeader Component > matches snapshot for completed task without due date 1`] = `
<DocumentFragment>
  <div
    style="display: flex; justify-content: space-between; margin-bottom: 16px;"
  >
    <button
      data-testid="ant-button"
      style="background-color: rgb(17, 164, 54); border: 1px solid rgb(17, 164, 54); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
      type="default"
    >
      <span
        style="font-size: 0.875rem; color: rgb(255, 255, 255);"
      >
        Completed
      </span>
    </button>
    <button
      data-testid="ant-button"
      style="background-color: rgb(245, 245, 245); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
      type="default"
    >
      <img
        alt="leading-icon"
        src="test-leading-icon.png"
      />
      <span
        style="font-size: 0.875rem;"
      >
        2023-12-31T00:00:00
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`TaskHeader Component > matches snapshot for incomplete task with due date 1`] = `
<DocumentFragment>
  <div
    style="display: flex; justify-content: space-between; margin-bottom: 16px;"
  >
    <button
      data-testid="ant-button"
      style="background-color: rgb(25, 107, 229); border: 1px solid rgb(25, 107, 229); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
      type="default"
    >
      <span
        style="font-size: 0.875rem; color: rgb(255, 255, 255);"
      >
        In Progress
      </span>
    </button>
    <button
      data-testid="ant-button"
      style="background-color: rgb(245, 245, 245); border-radius: 25px; padding: 0px 8px; display: flex; align-items: center;"
      type="default"
    >
      <img
        alt="leading-icon"
        src="test-leading-icon.png"
      />
      <span
        style="font-size: 0.875rem;"
      >
        2023-12-31T00:00:00
      </span>
    </button>
  </div>
</DocumentFragment>
`;
